Add AWS deployment scripts

lower29 opened the pull request:
**UPDATE: This has evolved over the past few months and so is more than the original pull request was aiming for. See my latest comment below for better description of the work/change that this introduces.**

I've added a `provision-and-deploy.sh` script that basically carries out all of the steps that were previously described in the README, but does them automatically. The only thing you have to do is authenticate with AWS (which the script helps with too).

The script attempts to:
1. Get an active AWS session to use, challenging the user and creating a new one if there isn't one, including MFA auth
1. Determine the status of the target EC2 instance for Docker Machine to use, creating (provisioning) one if it doesn't exist, or starting it if it is stopped.
1. Copy the required source files to the EC2 instance and running `make all SHIBBOLETH_CONFIG=archivematica` to deploy the services to the remote docker host.

Credentials from the standard `AWS_` environment variables are used where appropriate. Additional parameters are stored in `.machine-credentials`.

There are some configuration params that can't be guessed. These must be put into the `machine.conf`, which is based on the template `machine.conf.template`. This template must be copied, editied and renamed before use.

Once you have the template in place, just run `./provision-and-deploy.sh`, like the updated README says.

This is still a work in progress; it works for me but I haven't had chance to test it thoroughly. The intention was to make deploying to AWS extremely simple and repeatable.

This change builds on the changes in PR #26.

sevein commented:
This looks really good, @lower29. I have nothing to add but I've added @mamedin to the list of reviewers - I'm hoping he can add some feedback and/or do some testing.

sevein commented:
Is this still WIP, @lower29? I've added the WIP label instead of prefixing the title.

lower29 commented:
@sevein No, not really- I only put it as WIP in case it needed major work in response to comments. I'm happy for it to be merged as-is.

sevein commented:
@lower29 - this looks like it needs to be rebased but LGTM!

lower29 commented:
@sevein This has now been rebased... but I've also made some major changes to it too.

The new version expands its remit slightly, so the script now includes the provision and deployment of the Arkivum appliance and the NFS server, as well as the Docker host that is used by docker machine. This brings it more in line with what Jisc are deploying using their Terraform-based process.

There is still work to do on this; I'm passing it to @jeremysparks86 to pick up so that he can hopefully get the Arkivum appliance working and "kick the tyres" of the rest of it. Once that's happened then it'll hopefully be in a state where it can be reviewed formally and we can, fingers crossed, start using it to deploy our QA and demo environments in an automated and managed way.

lower29 commented:
Changed back to work-in-progress until we have got Arkivum appliance deployment working and the rest is working in a stable manner.

sevein commented on aws/machine/README.md:
```diff
@@ -38,105 +30,66 @@ Also, create a new SSH key pair:
+## Provisioning and Deployment
 
-Set up the client to connect to the machine we've just created:
+The [deploy](eploy.sh) script will automatically provision EC2 instances for an Arkivum appliance, a NFS server and a Docker host. The Docker host is provisioned using `docker-machine`, other instances and resources are provisioned using the AWS CLI tool. As well as provisioning, the deploy script also builds Docker images and mounts volumes to allow services to integrate and interact. The Docker ...
```
I think there's a typo in the link: should read `[deploy](deploy.sh)`.

lower29 replied:
Good spot. Will fix that.

lower29 commented:
I've added DNS records for each service deployed, as well as restructuring some of the code to make it easier to understand/maintain. I've also updated the readme, so that hopefully someone other than me can use this!

lower29 commented:
Rebased against current `master` branch.

Also:
* Made Shibboleth authentication optional for deployments
* Made Arkivum appliance optional for deployments
* Allowed Dockerhost root volume size to be configured as part of deployments
* Tidied code a bit and fixed some handling of errors when AWS account limits are hit (e.g. total size of all volumes)

lower29 commented:
I've now cleaned up the branch for this PR, squashing the 26-or-so commits into one. So what have we ended up with?

The `deploy.sh` and `teardown.sh` scripts together implement a consistent process for deployment and teardown of resources for RDSSARK into an AWS account. This is implemented using Bash and utilises the AWS CLI and Docker Machine to deploy instances, containers and other associated resources.

The scripts are configured using the `deployment.conf` configuration file, which can also be overriden with environment variables. For example, the `PROJECT_ID` and `ENVIRONMENT` parameters can be set as part of the command line, for example `PROJECT_ID=rl ENVIRONMENT=qa ./deploy.sh`. This allows multiple deployments to be managed using these scripts, enabling different environments to be deployed for different purposes - even if it's only a quick test session or demo.

Of paramount importance is the deployment's use of AWS resource tags. Where possible, all deployed resources are tagged with the `ProjectId` and `Environment` so that these attributes can be used to manage resources and track billing. The teardown script uses these tags to identify resources that need to be removed, avoiding dangling resources from being orphaned.

A complete deployment includes AWS instances for an Arkivum appliance, a Docker host running the full Docker Compose service set (Archivematica, NextCloud and RDSS channel adapter), and an NFS server for shared storage. The Arkivum appliance is optional, as is the use of Shibboleth for authentication. These and other options can be configured in the config file and via environment variables.

I've now run this up and down dozens of times, and have done a few end-to-end tests to confirm files can be added to NextCloud and processed using Archivematica.

@sevein and @mamedin, I know I've already requested reviews from you both in the past, but can you give this one more look through and let me know what you think? Thank you.

sevein commented on .travis.yml:
```diff
@@ -13,6 +13,8 @@ script:
   - python -myamllint -f parsable *-playbook.yml
   - ansible-lint -p *-playbook.yml
+  # Validate the shell scripts for aws/machine
+  - pushd aws/machine ; find . \( -name \*.sh -or -name \*.bash \) -exec shellcheck -x -f gcc {} + ; popd
```
Nice!

sevein commented on aws/machine/README.md:
```diff
@@ -26,6 +18,15 @@ You need:
 
+However, before attempting any deployment you will need the following:
+
+1. Acess to an AWS account (root or IAM credentials are fine)
```
Typo.

lower29 replied:
I forgot to fix this! I'll add a fixup as part of another PR.

mamedin commented:
Excellent work Richard. I have deployed a new QA env in less than 1 hour, it is great!

I would only point out these small suggestions:

- The default `NFS_STORAGE_VOLUME_TYPE` should be `gp2` to use the default nfs size.
- The default `DOCKERHOST_INSTANCE_TYPE` is too small, you get OOM errors when trying to deploy. I think that the default size should be "t2.large"
- After running the teardown.sh script, the route 53 hosted zones have to be deleted manually. I think that it should be mentioned on the "# Known Issues / TODO".

lower29 commented:
Thanks @mamedin for the useful feedback.

* I was testing `NFS_STORAGE_VOLUME_TYPE` locally with `sc1` and `NFS_STORAGE_SIZE` set to `500GB` - I didn't realise that you couldn't have less than 500GB for `st1` storage type. I'll change it to `gp2` with a note about changing the type if the size goes over 500.
* I've been using `DOCKERHOST_INSTANCE_TYPE` set to `t2.medium` locally, and that's working fine. I haven't been putting much work through the deployment with that setting, but it works, so I think that should be the minimum. I hadn't realised I'd forgotten to update the template config.
* I thought I'd gotten it so that all created resources were removed during teardown, including Route53 zones. I'll test again and if I find the resources are still left there then I'll add the known issue, like you suggest.

lower29 commented:
I've made the config changes that were suggested. I wasn't able to reproduce the Route53 zones not being removed so I left that as it was. If it turns out to be reproducable then we can raise an issue.

lower29 commented:
Merged! Thank you @sevein and @mamedin for all the feedback! 🙂

sloandw commented:
I have seen the same issue as Miguel in that the Route53 hosted zone wasn't removed when I ran the teardown script.
I had to manually remove the zone.
